src/topology/metric_space/infsep.lean

```diff
@@ -0,0 +1,369 @@
+end
+
+lemma le_infesep_image_injective_iff {d} {f : β → α} (hf : injective f) {s : finset β} :
+  d ≤ infesep (image f s) ↔ ∀ x y ∈ s, x ≠ y → d ≤ edist (f x) (f y) :=
+by simp_rw [le_infesep_image_iff, hf.ne_iff]
+
+lemma le_edist_of_le_infesep {d} (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) (hd : d ≤ s.infesep) :
+  d ≤ edist x y := le_infesep_iff.1 hd x hx y hy hxy
+
+lemma infesep_le_edist_of_mem (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) : s.infesep ≤ edist x y :=
+le_edist_of_le_infesep hx hy hxy le_rfl
+
+lemma le_infesep {d} (h : ∀ x y ∈ s, x ≠ y → d ≤ edist x y) : d ≤ s.infesep := le_infesep_iff.2 h
+
+lemma infesep_lt_iff {d} : s.infesep < d ↔ ∃ x y ∈ s, x ≠ y ∧ edist x y < d :=
+begin
+  simp_rw [infesep, finset.inf_lt_iff, exists_prop, mem_off_diag], split,
+  { rintros ⟨_, ⟨hx, hy, hxy⟩, hxyd⟩, exact ⟨_, hx, _, hy, hxy, hxyd⟩ },
+  { rintros ⟨x, hx, y, hy, hxy, hxyd⟩, exact ⟨(x, y), ⟨hx, hy, hxy⟩, hxyd⟩ }
+end
+
+lemma infesep_subsingleton (hs : ∀ {a b : α}, a ∈ s → b ∈ s → a = b) : s.infesep = ∞ :=
+eq_top_iff.2 (le_infesep (λ x hx y hy hxy, false.elim (hxy (hs hx hy))))
+
+@[simp] lemma infesep_empty : infesep (∅ : finset α) = ∞ :=
+infesep_subsingleton (λ _ _ ha _, false.elim (not_mem_empty _ ha))
+
+@[simp] lemma infesep_singleton : infesep ({x} : finset α) = ∞ :=
+infesep_subsingleton (λ a b ha hb, by rw [mem_singleton.1 ha, mem_singleton.1 hb])
+
+lemma infesep_anti : @antitone (finset α) _ _ _ infesep :=
+λ s t h, le_infesep $ λ x hx y hy, infesep_le_edist_of_mem (h hx) (h hy)
+
+lemma infesep_insert_le : infesep (insert x s) ≤ ⨅ (y ∈ s) (hxy : x ≠ y), edist x y :=
+begin
+  simp_rw [le_infi_iff],
+  refine λ _ hy hxy, infesep_le_edist_of_mem (mem_insert_self _ _) (mem_insert_of_mem hy) hxy
+end
+
+lemma infesep_pair_le_left (hxy : x ≠ y) : infesep ({x, y} : finset α) ≤ edist x y :=
+infesep_le_edist_of_mem (mem_insert_self _ _) (mem_insert_of_mem (mem_singleton_self _)) hxy
+
+lemma infesep_pair_le_right (hxy : x ≠ y) : infesep ({x, y} : finset α) ≤ edist y x :=
+by rw pair_comm; exact infesep_pair_le_left hxy.symm
+
+lemma le_infesep_pair : (edist x y) ⊓ (edist y x) ≤ infesep ({x, y} : finset α) :=
+begin
+  simp_rw [le_infesep_iff, inf_le_iff, mem_insert, mem_singleton],
+  rintros a (rfl | rfl) b (rfl | rfl) hab; finish
+end
+
+lemma infesep_pair_eq' (hxy : x ≠ y) : infesep ({x, y} : finset α) = (edist x y) ⊓ (edist y x) :=
+le_antisymm (_root_.le_inf (infesep_pair_le_left hxy) (infesep_pair_le_right hxy)) le_infesep_pair
+
+lemma infesep_exists (h : ∃ x y ∈ s, x ≠ y) : ∃ (x y ∈ s) (h : x ≠ y), edist x y = s.infesep :=
+begin
+  rw ← off_diag_nonempty_iff at h,
+  rcases exists_mem_eq_inf _ h (uncurry edist) with ⟨⟨x, y⟩, hxy, spec⟩,
+  rw mem_off_diag at hxy, rcases hxy with ⟨hx, hy, hxy⟩,
+  exact ⟨x, hx, y, hy, hxy, spec.symm⟩
+end
+
+lemma infesep_exists' (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y)
+  : ∃ (x y ∈ s) (h : x ≠ y), edist x y = s.infesep := infesep_exists ⟨x, hx, y, hy, hxy⟩
```

Comment:
There's no point having the first lemma, the assumption is that the caller should unpack existentials / conjunctions etc

Reply:
The only reason you might is that this is equivalent to `\not s.subsingleton`, or at least it is if we defined that. It's quite natural to have that as a condition on stuff. But I take your point - maybe for other things, but not this one?

Reply:
Also, have it as one condition makes it easier to use `off_diag_nonempty_iff`. So I think there is value in both...